clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareFetchRequestManager.java

```diff
@@ -34,30 +54,221 @@ public class ShareFetchRequestManager implements RequestManager {
+        long currentTimeMs = time.milliseconds();
+        Map<String, Uuid> topicIds = metadata.topicIds();
+
+        for (TopicPartition partition : fetchablePartitions()) {
+            SubscriptionState.FetchPosition position = subscriptions.position(partition);
+
+            if (position == null)
+                throw new IllegalStateException("Missing position for fetchable partition " + partition);
+
+            Optional<Node> leaderOpt = position.currentLeader.leader;
+
+            if (!leaderOpt.isPresent()) {
+                log.debug("Requesting metadata update for partition {} since the position {} is missing the current leader node", partition, position);
+                metadata.requestUpdate(false);
+                continue;
+            }
+
+            Node node = leaderOpt.get();
+            if (nodesWithPendingFetchRequests.contains(node.id())) {
+                log.trace("Skipping fetch for partition {} because previous fetch request to {} has not been processed", partition, node);
+            } else {
+                // if there is a leader and no in-flight requests, issue a new fetch
+                ShareSessionHandler.Builder builder = fetchable.computeIfAbsent(node, k -> {
+                    ShareSessionHandler shareSessionHandler = sessionHandlers.computeIfAbsent(node.id(), n -> new ShareSessionHandler(logContext, n, Uuid.randomUuid()));
+                    return shareSessionHandler.newBuilder();
+                });
+                Uuid topicId = topicIds.getOrDefault(partition.topic(), Uuid.ZERO_UUID);
+                builder.add(new TopicIdPartition(topicId, partition));
+
+                log.debug("Added {} fetch request for partition {} at position {} to node {}", fetchConfig.isolationLevel,
+                        partition, position, node);
+            }
+        }
+
+        Map<Node, ShareSessionHandler.ShareFetchRequestData> reqs = new LinkedHashMap<>();
+        for (Map.Entry<Node, ShareSessionHandler.Builder> entry : fetchable.entrySet()) {
+            reqs.put(entry.getKey(), entry.getValue().build());
+        }
+
+        // Because ShareFetcher doesn't have the concept of a validated position, we need to
+        // do some fakery here to keep the underlying fetching state management happy. This
+        // is a sign of a future refactor.
+        if (reqs.isEmpty()) {
+            for (TopicPartition tp : subscriptions.initializingPartitions()) {
+                Optional<Node> leader = metadata.currentLeader(tp).leader;
+                if (leader.isPresent()) {
+                    subscriptions.seekValidated(tp, new SubscriptionState.FetchPosition(0, Optional.empty(), metadata.currentLeader(tp)));
+                    subscriptions.maybeValidatePositionForCurrentLeader(new ApiVersions(), tp, metadata.currentLeader(tp));
+                }
+            }
+        }
+
+        return reqs;
+    }
+
+    private List<TopicPartition> fetchablePartitions() {
```

Comment:
This needs converting to TopicIdPartition. It's just another piece of the puzzle that we need to build.